File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public boolean solution(String[] phoneBook) {
        Arrays.sort(phoneBook, (a, b) -> Integer.compare(a.length(), b.length()));
        Set<String> prefixes = new HashSet<>();
        for (String phone : phoneBook) {
            for (int end = 1; end < phone.length(); end++) {
                if(prefixes.contains(phone.substring(0, end))) return false;
            }
            prefixes.add(phone);
        }

        return true;
    }
}

/*
가능한 모든 접두어를 구하고 확인하기
맵 ---- 낮은애들 부터 
*/

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    phone_book.sort()
    for i in range(1, len(phone_book)):
        if phone_book[i].startswith(phone_book[i-1]):
            return False
    return True
        
"""
폰북의 번호를 돌면서
번호를 정렬한다

1. 현재가 뒤의 있는 번호의 접두어이면 
    return False
2. 아니면
    return True


"""